Fix swap_tensors to swap PyObjects associated with TensorImpl

mikaylagawarecki opened the pull request:
This PR intends to fix the following issue when swapping two tensors

```python
>>> import torch
>>> torch.manual_seed(5)
>>> t1 = torch.randn(2)
>>> t2 = torch.randn(3)
>>> t1
tensor([-0.4868, -0.6038])
>>> t2
tensor([-0.5581,  0.6675, -0.1974])
>>> torch.utils.swap_tensors(t1, t2)
>>> t1
tensor([-0.5581,  0.6675, -0.1974])
>>> t2
tensor([-0.4868, -0.6038])
>>> t1.fill_(0.5) # t1 back to its unswapped state :o
tensor([-0.4868, -0.6038])
```

What happens here is that in `THPVariable_Wrap` (which is used when going back from C++ --> Python), we check if the TensorImpl of the tensor to be returned already has a pointer to a PyObject in its PyObject slot. If this is the case then this object is returned.

https://github.com/pytorch/pytorch/blob/57491d2046b7ddcdf5588dea3cc1dee6c1eb366b/torch/csrc/autograd/python_variable.cpp#L271-L292

When we run any operation that returns the same TensorImpl (e.g. inplace op, `t.to(dtype=t.dtype)`, etc.), although `t1` now has `t2`'s TensorImpl, `t2`'s TensorImpl still has a reference to `t2`, so when we do the op on `t1` and `THPVariable_Wrap` attempts to return the pointer to the TensorImpl's PyObject, we return a pointer to `t2` instead.

The TensorImpl should have the PyObjects in their PyObjectSlots swapped as well in `swap_tensors`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117913
* #118028
* #118023
* #117167
* __->__ #116955

mikaylagawarecki committed:
Fix swap_tensor to swap PyObjects associated with TensorImpl

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix swap_tensors to swap PyObjects associated with TensorImpl"

This PR intends to fix the following issue when swapping two tensors

```python
>>> import torch
>>> torch.manual_seed(5)
>>> t1 = torch.randn(2)
>>> t2 = torch.randn(3)
>>> t1
tensor([-0.4868, -0.6038])
>>> t2
tensor([-0.5581,  0.6675, -0.1974])
>>> torch.utils.swap_tensors(t1, t2)
>>> t1
tensor([-0.5581,  0.6675, -0.1974])
>>> t2
tensor([-0.4868, -0.6038])
>>> t1.fill_(0.5) # t1 is t2 again :o
tensor([-0.4868, -0.6038])
```

What happens here is that in `THPVariable_Wrap` (which is used when going back from C++ --> Python), we check if the TensorImpl of the tensor to be returned already has a pointer to a PyObject in its PyObject slot. If this is the case then this object is returned.

https://github.com/pytorch/pytorch/blob/57491d2046b7ddcdf5588dea3cc1dee6c1eb366b/torch/csrc/autograd/python_variable.cpp#L271-L292

When we run any operation that returns the same TensorImpl (e.g. inplace op, `t.to(dtype=t.dtype)`, etc.), although t1 now has `t2`'s TensorImpl, `t2`'s TensorImpl still has a reference to `t2`, so when we do the op on `t1` and `THPVariable_Wrap` attempts to return the pointer to the TensorImpl's PyObject, we return a pointer to `t2` instead.

The TensorImpl should have the PyObjects in their PyObjectSlots swapped as well in `swap_tensors`




[ghstack-poisoned]

albanD commented on torch/csrc/Module.cpp:
```diff
@@ -348,9 +348,41 @@ PyObject* THPModule_swap_tensor_impl(PyObject* _unused, PyObject* args) {
+
+  // The TensorImpls contain PyObjectSlots that have a reference to the PyObject
+  // associated with the TensorImpl. Swap this field as well.
+  c10::optional<PyObject*> mb_obj_a =
```
You could just use `a_` here. that should be the same object.
If we want to be extra safe, you can keep this and add another assert that mb_obj_a == a_.

albanD commented on torch/csrc/Module.cpp:
```diff
@@ -348,9 +348,41 @@ PyObject* THPModule_swap_tensor_impl(PyObject* _unused, PyObject* args) {
+  // The TensorImpls contain PyObjectSlots that have a reference to the PyObject
+  // associated with the TensorImpl. Swap this field as well.
+  c10::optional<PyObject*> mb_obj_a =
+      at::get_tensor_base(*(a->cdata))
+          .unsafeGetTensorImpl()
```
arent you able to do `a->cdata->unsafeGetTensorImpl()` directly?

albanD commented on torch/csrc/Module.cpp:
```diff
@@ -348,9 +348,41 @@ PyObject* THPModule_swap_tensor_impl(PyObject* _unused, PyObject* args) {
 
   // Swap the Tensor Impl
   c10::MaybeOwned<at::Tensor> tmp = a->cdata;
+
```
I just realized that we wanted to do other checks that I forgot to add here, could you add a check for both at::Tensors that a->cdata->use_count() == 1 (the one held by the PyObject) and a->cdata->weak_use_count() == 0.

albanD reviewed (commented):
Very good catch, thanks!

mikaylagawarecki committed:
Update on "Fix swap_tensors to swap PyObjects associated with TensorImpl"

This PR intends to fix the following issue when swapping two tensors

```python
>>> import torch
>>> torch.manual_seed(5)
>>> t1 = torch.randn(2)
>>> t2 = torch.randn(3)
>>> t1
tensor([-0.4868, -0.6038])
>>> t2
tensor([-0.5581,  0.6675, -0.1974])
>>> torch.utils.swap_tensors(t1, t2)
>>> t1
tensor([-0.5581,  0.6675, -0.1974])
>>> t2
tensor([-0.4868, -0.6038])
>>> t1.fill_(0.5) # t1 back to its unswapped state :o
tensor([-0.4868, -0.6038])
```

What happens here is that in `THPVariable_Wrap` (which is used when going back from C++ --> Python), we check if the TensorImpl of the tensor to be returned already has a pointer to a PyObject in its PyObject slot. If this is the case then this object is returned.

https://github.com/pytorch/pytorch/blob/57491d2046b7ddcdf5588dea3cc1dee6c1eb366b/torch/csrc/autograd/python_variable.cpp#L271-L292

When we run any operation that returns the same TensorImpl (e.g. inplace op, `t.to(dtype=t.dtype)`, etc.), although `t1` now has `t2`'s TensorImpl, `t2`'s TensorImpl still has a reference to `t2`, so when we do the op on `t1` and `THPVariable_Wrap` attempts to return the pointer to the TensorImpl's PyObject, we return a pointer to `t2` instead.

The TensorImpl should have the PyObjects in their PyObjectSlots swapped as well in `swap_tensors`




[ghstack-poisoned]

albanD reviewed (commented):
SGTM!
Please add a test!

mikaylagawarecki commented on test/test_torch.py:
```diff
@@ -9742,6 +9742,14 @@ def _checked_swap(self, t1, t2):
+        # returns doesn't change the TensorImpl, the returned value (which is
+        # given by returning the reference to the PyObject in the TensorImpl's
+        # PyObjectSlot) is still correct
+        self.assertEqual(id(t1.fill_(0.5)), id(t1))
```
@albanD this is tested here! I'm not sure how to create a Tensor from Python where TensorImpl's `use_count > 1` or `weak_use_count > 1` to test that the checks are exercised though, do you know

mikaylagawarecki committed:
Update on "Fix swap_tensors to swap PyObjects associated with TensorImpl"

This PR intends to fix the following issue when swapping two tensors

```python
>>> import torch
>>> torch.manual_seed(5)
>>> t1 = torch.randn(2)
>>> t2 = torch.randn(3)
>>> t1
tensor([-0.4868, -0.6038])
>>> t2
tensor([-0.5581,  0.6675, -0.1974])
>>> torch.utils.swap_tensors(t1, t2)
>>> t1
tensor([-0.5581,  0.6675, -0.1974])
>>> t2
tensor([-0.4868, -0.6038])
>>> t1.fill_(0.5) # t1 back to its unswapped state :o
tensor([-0.4868, -0.6038])
```

What happens here is that in `THPVariable_Wrap` (which is used when going back from C++ --> Python), we check if the TensorImpl of the tensor to be returned already has a pointer to a PyObject in its PyObject slot. If this is the case then this object is returned.

https://github.com/pytorch/pytorch/blob/57491d2046b7ddcdf5588dea3cc1dee6c1eb366b/torch/csrc/autograd/python_variable.cpp#L271-L292

When we run any operation that returns the same TensorImpl (e.g. inplace op, `t.to(dtype=t.dtype)`, etc.), although `t1` now has `t2`'s TensorImpl, `t2`'s TensorImpl still has a reference to `t2`, so when we do the op on `t1` and `THPVariable_Wrap` attempts to return the pointer to the TensorImpl's PyObject, we return a pointer to `t2` instead.

The TensorImpl should have the PyObjects in their PyObjectSlots swapped as well in `swap_tensors`




[ghstack-poisoned]

mikaylagawarecki commented on torch/csrc/Module.cpp:
```diff
@@ -346,11 +346,49 @@ PyObject* THPModule_swap_tensor_impl(PyObject* _unused, PyObject* args) {
+      "Expected single reference to b's Tensor object but got ",
+      b->cdata->use_count());
+  // weak_use_count() adds 1 if use_count is non-zero
+  TORCH_CHECK(
```
There is  a weak_ref check  on the Pyobj, do `a->cdata->weak_use_count()` and `weakref.getweakrefs(t)` differ?

I am thinking the answer is yes because `a->cdata->weak_use_count()` probably counts cpp and python weakrefs while I think `weakref.getweakrefs(t)` would only count the python ones.

Is there a way to install a cpp-only weakref to exercise this check?

mikaylagawarecki committed:
Update on "Fix swap_tensors to swap PyObjects associated with TensorImpl"

This PR intends to fix the following issue when swapping two tensors

```python
>>> import torch
>>> torch.manual_seed(5)
>>> t1 = torch.randn(2)
>>> t2 = torch.randn(3)
>>> t1
tensor([-0.4868, -0.6038])
>>> t2
tensor([-0.5581,  0.6675, -0.1974])
>>> torch.utils.swap_tensors(t1, t2)
>>> t1
tensor([-0.5581,  0.6675, -0.1974])
>>> t2
tensor([-0.4868, -0.6038])
>>> t1.fill_(0.5) # t1 back to its unswapped state :o
tensor([-0.4868, -0.6038])
```

What happens here is that in `THPVariable_Wrap` (which is used when going back from C++ --> Python), we check if the TensorImpl of the tensor to be returned already has a pointer to a PyObject in its PyObject slot. If this is the case then this object is returned.

https://github.com/pytorch/pytorch/blob/57491d2046b7ddcdf5588dea3cc1dee6c1eb366b/torch/csrc/autograd/python_variable.cpp#L271-L292

When we run any operation that returns the same TensorImpl (e.g. inplace op, `t.to(dtype=t.dtype)`, etc.), although `t1` now has `t2`'s TensorImpl, `t2`'s TensorImpl still has a reference to `t2`, so when we do the op on `t1` and `THPVariable_Wrap` attempts to return the pointer to the TensorImpl's PyObject, we return a pointer to `t2` instead.

The TensorImpl should have the PyObjects in their PyObjectSlots swapped as well in `swap_tensors`




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix swap_tensors to swap PyObjects associated with TensorImpl"

This PR intends to fix the following issue when swapping two tensors

```python
>>> import torch
>>> torch.manual_seed(5)
>>> t1 = torch.randn(2)
>>> t2 = torch.randn(3)
>>> t1
tensor([-0.4868, -0.6038])
>>> t2
tensor([-0.5581,  0.6675, -0.1974])
>>> torch.utils.swap_tensors(t1, t2)
>>> t1
tensor([-0.5581,  0.6675, -0.1974])
>>> t2
tensor([-0.4868, -0.6038])
>>> t1.fill_(0.5) # t1 back to its unswapped state :o
tensor([-0.4868, -0.6038])
```

What happens here is that in `THPVariable_Wrap` (which is used when going back from C++ --> Python), we check if the TensorImpl of the tensor to be returned already has a pointer to a PyObject in its PyObject slot. If this is the case then this object is returned.

https://github.com/pytorch/pytorch/blob/57491d2046b7ddcdf5588dea3cc1dee6c1eb366b/torch/csrc/autograd/python_variable.cpp#L271-L292

When we run any operation that returns the same TensorImpl (e.g. inplace op, `t.to(dtype=t.dtype)`, etc.), although `t1` now has `t2`'s TensorImpl, `t2`'s TensorImpl still has a reference to `t2`, so when we do the op on `t1` and `THPVariable_Wrap` attempts to return the pointer to the TensorImpl's PyObject, we return a pointer to `t2` instead.

The TensorImpl should have the PyObjects in their PyObjectSlots swapped as well in `swap_tensors`




[ghstack-poisoned]

albanD commented on test/test_torch.py:
```diff
@@ -9763,10 +9771,18 @@ def test_swap_basic(self):
+                t3 = t1.clone().detach().requires_grad_(True)
+                out = t3 * 2
+                with self.assertRaisesRegex(RuntimeError, "Expected single reference to a's"):
+                    torch.utils.swap_tensors(t3, t2)
```
nit: can you add
```python
del out
# Now succeeds
torch.utils.swap_tensors(t3, t2)
```

albanD reviewed (approved):
Small addition to the test, But sounds good otherwise

mikaylagawarecki committed:
Update on "Fix swap_tensors to swap PyObjects associated with TensorImpl"

This PR intends to fix the following issue when swapping two tensors

```python
>>> import torch
>>> torch.manual_seed(5)
>>> t1 = torch.randn(2)
>>> t2 = torch.randn(3)
>>> t1
tensor([-0.4868, -0.6038])
>>> t2
tensor([-0.5581,  0.6675, -0.1974])
>>> torch.utils.swap_tensors(t1, t2)
>>> t1
tensor([-0.5581,  0.6675, -0.1974])
>>> t2
tensor([-0.4868, -0.6038])
>>> t1.fill_(0.5) # t1 back to its unswapped state :o
tensor([-0.4868, -0.6038])
```

What happens here is that in `THPVariable_Wrap` (which is used when going back from C++ --> Python), we check if the TensorImpl of the tensor to be returned already has a pointer to a PyObject in its PyObject slot. If this is the case then this object is returned.

https://github.com/pytorch/pytorch/blob/57491d2046b7ddcdf5588dea3cc1dee6c1eb366b/torch/csrc/autograd/python_variable.cpp#L271-L292

When we run any operation that returns the same TensorImpl (e.g. inplace op, `t.to(dtype=t.dtype)`, etc.), although `t1` now has `t2`'s TensorImpl, `t2`'s TensorImpl still has a reference to `t2`, so when we do the op on `t1` and `THPVariable_Wrap` attempts to return the pointer to the TensorImpl's PyObject, we return a pointer to `t2` instead.

The TensorImpl should have the PyObjects in their PyObjectSlots swapped as well in `swap_tensors`




[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge

mikaylagawarecki commented:
@pytorchbot merge
